### ISSUE TITLE
Prototype a JSON API using ESP8266 Async HTTP server

### ISSUE THREAD

**Issue by 3rwww1** [LeClubSandwichStudio/LaCOOLBoard#77]
**timebox:** 2d

**Todo**

#### get the current WiFi status
request: `GET /wifi/status`
response code: `200 OK`
response body:
```json
{
  "state": "CONNECTED",
  "network" : {
    "ssid": "lapaillasse",
    "bssid": "24:a4:3c:9e:d2:84",
    "rssi": -67,
    "security": "WPA2_PSK",
  }
}

```
#### scan WiFi networks
request URI: `PUT /wifi/scan`
response code: `200 OK`
response:
```json
[{
  "ssid": "lapaillasse",
  "bssid": "24:a4:3c:9e:d2:84",
  "rssi": -67,
  "security": "WPA2-PSK"
}, {
  "ssid": "erwan-home",
  "bssid": "4f:a6:3c:9c:ff:57",
  "rssi": -67,
  "security": "WPA2-PSK"
}]
```

#### list all saved networks
request URI: `GET /wifi/saved/`
response code: `200 OK`
response body:
```json
[{
  "ssid": "erwan-home",
  "bssid": "4f:a6:3c:9c:ff:57",
  "security": "WPA2_PSK",
  "psk": "biohacking",
  "rssi": -73
}, {
  "ssid": "lapaillasse",
  "bssid": "24:a4:3c:9e:d2:84",
  "security": "WPA2-PSK",
  "psk": "va-te-faire-connecter",
  "rssi": -67,
}]
```
#### save a new wifi network to the board
request: `POST /wifi/saved`
request body:
```json
{
  "ssid": "erwan-home",
  "bssid": "4f:a6:3c:9c:ff:57",
  "psk": "va-te-faire-connecter"
}
```
response code: `201 CREATED`
response body:
```json
{
  "ssid": "erwan-home",
  "bssid": "24:a4:3c:9e:d2:84",
  "rssi": -67,
  "psk": "va-te-faire-connecter"
}
```

#### forget all WiFi networks
request uri: `DELETE /wifi/saved`
response code: `204 NO CONTENT`

#### forget one WiFi network
request uri: `DELETE /wifi/saved:bssid`
response code: `204 NO CONTENT`
(edited)

#### connect to a saved wifi network
request uri: `PUT /wifi/saved/:bssid/connect`
response code: `202 ACCEPTED`
response body:
```
{
  "attemptId": "0fe239a4-ed8d-49e9-9838-78e764cb111e"
}
```

#### SSE
see [this link for reference](https://www.html5rocks.com/en/tutorials/eventsource/basics/) or [this one at Mozilla](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events)
SSE endpoint: `/sse`
```
event: wifiConnectionResult
id: 0fe239a4-ed8d-49e9-9838-78e764cb111e
data: {
data:   "outcome": "FAILURE",
data:   "reason": "AUTH_FAILED"
data: }
```

https://github.com/me-no-dev/ESPAsyncWebServer

**Comment by Kameeno** [comment]
is done, is available on: http://192.168.4.1/scan
he get an array like that: 
```
[
  {
    "rssi": -85,
    "ssid": "network0",
    "bssid": "00:00:00:00:00:00",
    "channel": 6,
    "secure": 8,
    "hidden": false
  },
  {
    "rssi": -47,
    "ssid": "network1",
    "bssid": "00:00:00:00:00:00",
    "channel": 6,
    "secure": 8,
    "hidden": false
  },
  {
    "rssi": -77,
    "ssid": "network2",
    "bssid": "00:00:00:00:00:00",
    "channel": 6,
    "secure": 8,
    "hidden": false
  },
  {
    "rssi": -68,
    "ssid": "network3",
    "bssid": "00:00:00:00:00:00",
    "channel": 11,
    "secure": 8,
    "hidden": false
  },
  {
    "rssi": -53,
    "ssid": "network4",
    "bssid": "00:00:00:00:00:00",
    "channel": 11,
    "secure": 7,
    "hidden": false
  },
  {
    "rssi": -71,
    "ssid": "network5",
    "bssid": "00:00:00:00:00:00",
    "channel": 11,
    "secure": 8,
    "hidden": false
  },
  {
    "rssi": -77,
    "ssid": "network6",
    "bssid": "00:00:00:00:00:00",
    "channel": 11,
    "secure": 8,
    "hidden": false
  }
]

```

**Comment by nlsdvl** [comment]
the task related to network service publishing/discovery could be split in another ticket (is it already ?)
I've been testing it today, the discovery worked with:
-- `gssdp-discovery -m all`
-- `dns-sd -B` 
I was able to access the device on the network with `http://coolboard-2c3ae84fbe5b.local/` at some point, but this did not work reliably at later attempts.

I am now looking for an appropriate package that can list the service in javascript.

regarding the API, I was able to use the following endpoints with Curl : 
/add/wifi
/reset/wifi
/wifi/scan

the API calls did not work with Postman however.

**Comment by nlsdvl** [comment]
API calls from JS working fine.

**Comment by nlsdvl** [comment]
service discovery working great with:  https://github.com/taoyuan/nbonjour

**Comment by nlsdvl** [comment]
> request: POST /wifi/saved

There is no way to validate the password that the user posts without disconnecting. So a success response means the password is stored successfully. But not that it is a valid password for that network.

**Comment by nlsdvl** [comment]
How about PUT & DELETE ? We need to be able to update or delete WiFis ... current @Kameeno implemented a resetWiFi method for that purpose.

**Comment by nlsdvl** [comment]
Maybe not a priority, but there's no mention of a `/disconnect` command. I expect `/connect` implicitly disconnect if needed. Disconnect would be used to get back to AP mode, or for test/debug purpose.

cc @3rwww1 @Kameeno  you guys reading this ?

**Comment by nlsdvl** [comment]
I'd expect an error (4xx client error) when asking the board to forget to a Wifi but the board being actually connected to the same wifi. Correct ?

**Comment by Kameeno** [comment]
@nlsdvl  yes, i'm not arriving to reconnect to another wifiNetwork without crashing...
